staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -462,6 +462,9 @@ func (c *Cacher) Watch(ctx context.Context, key string, opts storage.ListOptions
 		return nil, err
 	}
 
+	if c.stopped {
+		return nil, storage.NewInternalError("cacher stopped")
```

Comment:
This is still racy, right?

Can we change that to something like:

1. use cacher.stopCh that we already have
2. pass that down to cacher.ready (via newReady)
3. newReady() starting a goroutine that will wait on that channel and on its closure switch ready to "stopped"
4. switch ready.wait() to instead of returning bool, additionally to be able to return an error (and the only error that it will be returning for now will be "stopped")
?

The above seems not visibly more complicated and non-racy.

Reply:
yeah, that sounds more correct , I was afraid of changing a conditional variable for that ...
since we are doing it, do we want to pass a context too?

Reply:
to wait() method? yeah probably

Reply:
heh https://github.com/golang/go/issues/16620

Reply:
I think that is easier to synchronize using only channels.